mbuild/conversion.py

```diff
@@ -0,0 +1,1364 @@
+                parent_compound = residue_compound
+            else:
+                parent_compound = chain_compound
+            for atom in residue.atoms:
+                pos = np.array([atom.xx,
+                                atom.xy,
+                                atom.xz]) / 10
+                new_atom = mb.Particle(name=str(atom.name),
+                                    pos=pos)
+                parent_compound.add(new_atom, label='{0}[$]'.format(atom.name))
+                atom_mapping[atom] = new_atom
+
+    # Infer bonds information
+    for bond in structure.bonds:
+        atom1 = atom_mapping[bond.atom1]
+        atom2 = atom_mapping[bond.atom2]
+        compound.add_bond((atom1, atom2))
+
+    # Convert box information
+    if structure.box is not None:
+        compound.periodicity = structure.box[0:3] / 10
+    else:
+        compound.periodicity = np.array([0., 0., 0.])
+
+    return compound
+
+def from_trajectory(traj, compound=None, frame=-1,
+                    coords_only=False, infer_hierarchy=True):
+    """Extract atoms and bonds from a md.Trajectory.
+
+    Will create sub-compounds for every chain if there is more
+    than one and sub-sub-compounds for every residue.
+
+    Parameters
+    ----------
+    traj : mdtraj.Trajectory
+        The trajectory to load.
+    compound : mb.Compound, optional, default=None
+        Host mb.Compound that we are loading to.
+    frame : int, optional, default=-1 (last)
+        The frame to take coordinates from.
+    coords_only : bool, optional, default=False
+        Only read coordinate information
+    infer_hierarchy : bool, optional, default=True
+        If True, infer compound hierarchy from chains and residues
+
+    Returns
+    -------
+    compound : mb.Compound
+
+    """
+    # Check for coords_only option
+    if compound and coords_only:
+        if traj.n_atoms != compound.n_particles:
+            raise ValueError('Number of atoms in {traj} does not match'
+                             ' {compound}'.format(**locals()))
+        atoms_particles = zip(traj.topology.atoms,
+                              compound.particles(include_ports=False))
+        if None in compound._particles(include_ports=False):
+            raise ValueError('Some particles are None')
+        for mdtraj_atom, particle in atoms_particles:
+            particle.pos = traj.xyz[frame, mdtraj_atom.index]
+        return compound
+    elif coords_only and not compound:
+        raise MBuildError('coords_only=True but'
+                 'host compound is not provided')
+
+    # Initialize a compound if none is provided
+    if not compound:
+        compound = mb.Compound()
+
+    atom_mapping = dict()
+    for chain in traj.topology.chains:
+        if traj.topology.n_chains > 1:
+            chain_compound = mb.Compound()
+            compound.add(chain_compound, 'chain[$]')
+        else:
+            chain_compound = compound
+        for res in chain.residues:
```

Comment:
Same comment about testing with a complex protein molecule to see how well the information is preserved. The best test for these would be a `from_mdtraj`, `to_mdtraj` sequence I suppose.